### PR TITLE
Add ephemeral storage support for JavaScript cookies

### PR DESCRIPTION
**Pull request by mrobinson**
This change adds a new class, EphemeralStoragePartition, which provides
access to an ephemeral version of a RestrictedCookieManager for the
content process. In an upcoming change this will also provide a
StoragePartition for navigation requests in third-party frames in order
to add support for navigation cookies.

Resolves brave/brave-browser#12591

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [x] Linux
  - [ ] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [x] Linux
  - [ ] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`, `npm run gn_check`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Requested a security/privacy review as needed.
- [x] Added appropriate QA labels (`QA/Yes` or `QA/No`) to the associated issue
- [x] Added appropriate release note labels (`release-notes/include` or `release-notes/exclude`) to the associated issue
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: patches/content-browser-renderer_host-render_frame_host_impl.cc.patch [hunk 1]
```diff
@@ -0,0 +1,12 @@
+diff --git a/content/browser/renderer_host/render_frame_host_impl.cc b/content/browser/renderer_host/render_frame_host_impl.cc
+index 497033d3d5257daffe4aca2ebdd1fd71513db73f..32ab8f258aa8f69e33854f411b39d510b885ecf3 100644
+--- a/content/browser/renderer_host/render_frame_host_impl.cc
++++ b/content/browser/renderer_host/render_frame_host_impl.cc
+@@ -7840,6 +7840,7 @@ void RenderFrameHostImpl::BindSmsReceiverReceiver(
+ 
+ void RenderFrameHostImpl::BindRestrictedCookieManager(
+     mojo::PendingReceiver<network::mojom::RestrictedCookieManager> receiver) {
++  BRAVE_BINDRESTRICTEDCOOKIEMANAGER
```
Comment (bridiver) [inline patches/content-browser-renderer_host-render_frame_host_impl.cc.patch:9]: this patch is not necessary because there is already a way in ContentBrowserClient to create a special restricted cookie manager, but I think handling JS and network cookies separately is a mistake. I think we need to pass the 3p cookie info into the network service cookie manager so we can handle them all in the same way

Comment (mrobinson) [reply patches/content-browser-renderer_host-render_frame_host_impl.cc.patch:9]: Our plan is to handle navigation cookies in a similar fashion, by implementing the abstract StoragePartition interface in `EphemeralStoragePartition`. Calls to the interface that do not access the `NetworkContext` will be routed to the default `StoragePartition`. Then we will use this `StoragePartition` when spinning up navigation in `NavigationRequest`. This work has bit-rotted a bit, but here is how this looks:

```diff
diff --git a/chromium_src/content/browser/renderer_host/navigation_request.cc b/chromium_src/content/browser/renderer_host/navigation_request.cc
index 635d7b75fd..5bf6d73bee 100644
--- a/chromium_src/content/browser/renderer_host/navigation_request.cc
+++ b/chromium_src/content/browser/renderer_host/navigation_request.cc
@@ -3,6 +3,7 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this file,
  * You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+#include "brave/content/browser/ephemeral_storage_partition.h"
 #include "content/browser/renderer_host/frame_tree.h"
 #include "content/browser/renderer_host/frame_tree_node.h"
 #include "url/gurl.h"
@@ -46,6 +47,18 @@ GURL GetTopDocumentGURL(content::FrameTreeNode* frame_tree_node) {
       GetTopDocumentGURL(frame_tree_node_), frame_tree_node_->IsMainFrame(), \
       common_params_->method, &common_params_->referrer);
 
+#define BRAVE_ONSTARTCHECKSCOMPLETE_REPLACESTORAGEPARTITION                  \
+  const GURL top_level_url = GetTopDocumentGURL(frame_tree_node_);           \
+  const url::Origin top_level_origin = url::Origin::Create(top_level_url);   \
+  const url::Origin target_origin = url::Origin::Create(common_params_->url);\
+  if (!navigating_frame_host->is_main_frame() &&                             \
+      !top_level_origin.IsSameOriginWith(target_origin)) {                   \
+    EphemeralStoragePartition* ephemeral_partition =                         \
+        browser_context->GetExistingEphemeralStoragePartition(top_level_url);\
+    if (ephemeral_partition)                                                 \
+      partition = ephemeral_partition;                                       \
+  }
+
 #include "../../../../../content/browser/renderer_host/navigation_request.cc"
```

File: chromium_src/content/public/browser/browser_context.h [hunk 2]
```diff
@@ -6,16 +6,17 @@
 #ifndef BRAVE_CHROMIUM_SRC_CONTENT_PUBLIC_BROWSER_BROWSER_CONTEXT_H_
 #define BRAVE_CHROMIUM_SRC_CONTENT_PUBLIC_BROWSER_BROWSER_CONTEXT_H_
 
-#include "../../../../../content/public/browser/browser_context.h"
-
 #include <string>
 
 #include "base/memory/ref_counted.h"
+#include "brave/content/browser/ephemeral_storage_partition.h"
```
Comment (mrobinson) [inline chromium_src/content/public/browser/browser_context.h:12]: This is an issue because we are including a non-public header. I will fix this in a new version of the branch.

File: build/config/brave_build.gni [hunk 3]
```diff
@@ -8,9 +8,10 @@ import("//brave/build/config/compiler.gni")
 import("//brave/build/features.gni")
 import("//brave/components/omnibox/browser/sources.gni")
 import("//brave/components/search_engines/sources.gni")
-import("//brave/components/sync/sources.gni")
 import("//brave/components/sync/driver/sources.gni")
+import("//brave/components/sync/sources.gni")
 import("//brave/components/sync_device_info/sources.gni")
+import("//brave/content/browser/sources.gni")
```
Comment (iefremov) [inline build/config/brave_build.gni:14]: wondering why this is needed

File: content/browser/sources.gni [hunk 4]
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2020 The Brave Authors. All rights reserved.
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+
+brave_content_browser_sources = [
+  "//brave/content/browser/ephemeral_storage_partition.cc",
```
Comment (iefremov) [inline content/browser/sources.gni:7]: since we ended up with adding new files to `content/browser`, is it possible to move code from `browser_context.cc/h` override to a new file? AFAIU that code is not really related to `BrowserContext`

Comment (iefremov) [reply content/browser/sources.gni:7]: E.g. `GetSessionStorageNamespaceId()`

File: content/browser/ephemeral_storage_partition.h [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+namespace content {
+
+using EphemeralStorageParitionMapKey =
+    std::pair<content::BrowserContext*, std::string>;
+
+class StoragePartitionImpl;
+
+// EphemeralStoragePartition manages an in-memory NetworkingContext to provide
+// ephemeral cookie storage.
+class EphemeralStoragePartition
```
Comment (iefremov) [inline content/browser/ephemeral_storage_partition.h:25]: If this is moved to `content/public` then we can avoid patching BrowserContext, right?

File: content/browser/ephemeral_storage_partition.h [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+#include <string>
+#include <utility>
+
+#include "content/public/browser/storage_partition.h"
+#include "services/network/public/mojom/network_context.mojom.h"
+#include "base/memory/ref_counted.h"
+
+namespace content {
+
+using EphemeralStorageParitionMapKey =
```
Comment (iefremov) [inline content/browser/ephemeral_storage_partition.h:18]: Parition -> Partition
